xip/xipether.c

```diff
@@ -215,6 +215,224 @@ static int do_delneigh(int argc, char **argv)
+
+	UNUSED(who);
+
+	if (n->nlmsg_type != RTM_NEWROUTE && n->nlmsg_type != RTM_DELROUTE) {
+		fprintf(stderr, "Not a route: %08x %08x %08x\n",
+			n->nlmsg_len, n->nlmsg_type, n->nlmsg_flags);
+		return 0;
+	}
+	if (r->rtm_family != AF_XIA) {
+		/* fprintf(stderr, "Wrong rtm_family %d\n", r->rtm_family); */
+		return 0;
+	}
+	len -= NLMSG_LENGTH(sizeof(*r));
+	if (len < 0) {
+		fprintf(stderr, "BUG: wrong nlmsg len %d\n", len);
+		return -1;
+	}
+	if (r->rtm_dst_len != sizeof(struct xia_xid)) {
+		fprintf(stderr, "BUG: wrong rtm_dst_len %d\n", r->rtm_dst_len);
+		return -1;
+	}
+
+	/* XXX Doesn't the kernel provide similar function? */
+	parse_rtattr(tb, RTA_MAX, RTM_RTA(r), len);
+	table = rtnl_get_table(r, tb);
+
+	/* Filter happens here. */
+	if (table != XRTABLE_MAIN_INDEX)
+		return 0;
+	if (!tb[RTA_DST] ||
+		RTA_PAYLOAD(tb[RTA_DST]) != sizeof(struct xia_xid) ||
+		r->rtm_dst_len != sizeof(struct xia_xid))
+		return -1;
+	dst = (const struct xia_xid *)RTA_DATA(tb[RTA_DST]);
+	if (dst->xid_type != filter.xid_type)
+		return 0;
+
+	if (n->nlmsg_type == RTM_DELROUTE)
+		fprintf(fp, "Deleted ");
+	fprintf(fp, "to ");
+	/* XXX It got to use @fp! */
+	print_xia_xid(dst);
+	fprintf(fp, "\n");
+
+	if (tb[RTA_MULTIPATH]) {
+		struct rtnl_xia_ether_addrs *rtha =
+			RTA_DATA(tb[RTA_MULTIPATH]);
+		int len = RTA_PAYLOAD(tb[RTA_MULTIPATH]);
+		char ha[MAX_ADDR_LEN];
+
+		if (RTHA_OK(rtha, len)) {
+			/* We only have a header, nothing else. */
+			assert(rtha->attr_len == sizeof(*rtha));
+
+			assert(!lladdr_ntop(rtha->interface_addr, rtha->interface_addr_len,
+				ha, sizeof(ha)));
+			fprintf(fp, "lladdr: %s\tdev: %s\n", ha,
+				ll_index_to_name(rtha->interface_index));
+		}
+	}
+
+	assert(!r->rtm_src_len);
+	/* XXX It should go to be printed out in flags. */
+	assert(!(r->rtm_flags & RTM_F_CLONED));
+
+	/* XXX This should become a function, and removed mixed flags, that is,
+	 * it doesn't make sense to have RTNH_F_* and RTM_F_* together.
+	 */
+	fprintf(fp, "flags [");
+	if (r->rtm_flags & RTNH_F_DEAD)
+		fprintf(fp, "dead ");
+	if (r->rtm_flags & RTNH_F_ONLINK)
+		fprintf(fp, "onlink ");
+	if (r->rtm_flags & RTNH_F_PERVASIVE)
+		fprintf(fp, "pervasive ");
+	if (r->rtm_flags & RTM_F_NOTIFY)
+		fprintf(fp, "notify ");
+	fprintf(fp, "]");
+
+	fprintf(fp, "\n\n");
+	fflush(fp);
+	return 0;
+}
+
+static int showinfo(rtnl_filter_t filter)
+{
+	reset_filter();
+
+	if (rtnl_wilddump_request(&rth, AF_XIA, RTM_GETROUTE) < 0) {
+		perror("Cannot send dump request");
+		exit(1);
+	}
+	if (rtnl_dump_filter(&rth, filter, stdout, NULL, NULL) < 0) {
+		fprintf(stderr, "Dump terminated\n");
+		exit(1);
+	}
+
+	return 0;
+}
+
+static int do_show(int argc, char **argv)
+{
+	const char *name;
+
+	if (argc != 1) {
+		fprintf(stderr, "Wrong number of parameters\n");
+		return usage();
+	}
+
+	name = argv[0];
+	if (!matches(name, "interfaces")) {
+		return showinfo(print_interface);
+	}
+	else if (!matches(name, "neighs")) {
+		return showinfo(print_neigh);
+	}
+	else {
```

Comment:
``else {`` should be put in the above line, that is, ``} else {``. Same for ``else if`` above.